### PR TITLE
Add target components handling in `get_level_dataframe`

### LINKED ISSUE #1169: Add target components handling in `get_level_dataframe`
### 🚀 Feature Request

Add target components handling in `get_level_dataframe`

### Proposal

1. Change the behaviour of `get_level_dataframe`, now it should transform all the columns in passed dataframe
2. In `BaseReconciliator.reconcile`
- Fix `get_level_dataframe` usage, passed dataframe should contain target, quantiles and target components
- Extract target components from dataframe after `get_level_dataframe`
- Add target components into resulting dataset in the end of the methods using `add_target_components`
4. In `TSDataset.get_level_dataset`
- Fix `get_level_dataframe` usage, passed dataframe should contain target, quantiles and target components
- `target_names` now should also contains target components from the dataset:
https://github.com/tinkoff-ai/etna/blob/ea1888e78ed1df030eae340eefe895456f30810a/etna/datasets/tsdataset.py#L1099
- Extract target components from `target_level_df`
- Add target components into resulting dataset in the end of the methods using `add_target_components` instead of:
https://github.com/tinkoff-ai/etna/blob/ea1888e78ed1df030eae340eefe895456f30810a/etna/datasets/tsdataset.py#L1109

### Test cases
1. For `get_level_dataframe`
- Fix the current test
- Check that it works correct in case with target components
2. For `BaseReconciliator.reconcile`
- Check that in case with target components the resulting dataset contains target components
3. For `TSDataset.get_level_dataset`
- Check that it works correct in case with target components(contains components + components are correct)
### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alex-hse-repository**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1169

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 ### Added
+- Add target components handling in get_level_dataframe ([#1179](https://github.com/tinkoff-ai/etna/pull/1179))
```
Comment (Mr-Geekman) [inline CHANGELOG.md:10]: get_level_dataframe -> `get_level_dataframe`

File: etna/datasets/tsdataset.py [hunk 2]
```diff
@@ -1111,21 +1113,29 @@ def get_level_dataset(self, target_level: str) -> "TSDataset":
         if target_level_index > current_level_index:
             raise ValueError("Target level should be higher in the hierarchy than the current level of dataframe!")
 
+        target_names = tuple(get_target_with_quantiles(columns=self.columns))
```
Comment (Mr-Geekman) [inline etna/datasets/tsdataset.py:1116]: Don't we want to make a function, that does this logic?

File: etna/datasets/tsdataset.py [hunk 3]
```diff
@@ -1111,21 +1113,29 @@ def get_level_dataset(self, target_level: str) -> "TSDataset":
...
                 target_level_segments=target_level_segments,
             )
 
         else:
-            target_names = tuple(get_target_with_quantiles(columns=self.columns))
-            target_level_df = self[:, current_level_segments, target_names]
+            target_level_df = self.to_pandas(features=target_names)
+
+        target_components_df = None
+        if self.target_components_names is not None:
```
Comment (Mr-Geekman) [inline etna/datasets/tsdataset.py:1136]: There are a lot of this check: `if self.target_components_names is not None:`, can we somehow avoid it?

File: etna/datasets/utils.py [hunk 4]
```diff
@@ -225,15 +225,11 @@ def get_level_dataframe(
     :
        dataframe at the target level
```
Comment (Mr-Geekman) [inline etna/datasets/utils.py:226]: At parameters section you write that `source_level_segments` is tuple, but in typing it is `List`.

File: etna/datasets/utils.py [hunk 5]
```diff
@@ -225,15 +225,11 @@ def get_level_dataframe(
...
-    num_target_names = len(target_names)
+    column_names = sorted(set(df.columns.get_level_values("feature")))
+    num_columns = len(column_names)
     num_source_level_segments = len(source_level_segments)
     num_target_level_segments = len(target_level_segments)
 
-    if len(target_names) == 0:
-        raise ValueError("Provided dataframe has no columns with the target variable!")
-
     if set(df.columns.get_level_values(level="segment")) != set(source_level_segments):
```
Comment (Mr-Geekman) [inline etna/datasets/utils.py:233]: Why do we need `source_level_segments` if we enforce `df.columns` to be compatible with them?

File: etna/datasets/utils.py [hunk 6]
```diff
@@ -243,30 +239,30 @@ def get_level_dataframe(
     if num_target_level_segments != mapping_matrix.shape[0]:
         raise ValueError("Number of target level segments do not match mapping matrix number of columns!")
 
-    df = df.loc[:, pd.IndexSlice[source_level_segments, target_names]]
-
-    source_level_data = df.values  # shape: (t, num_source_level_segments * num_target_names)
+    source_level_data = df.loc[
+        pd.IndexSlice[:], pd.IndexSlice[source_level_segments, column_names]
```
Comment (Mr-Geekman) [inline etna/datasets/utils.py:243]: Here we go again with this cursed indexing...

Comment (Mr-Geekman) [reply etna/datasets/utils.py:243]: May be we should mark it somehow to find it later.

File: etna/reconciliation/base.py [hunk 7]
```diff
@@ -83,18 +85,29 @@ def reconcile(self, ts: TSDataset) -> TSDataset:
         current_level_segments = ts.hierarchical_structure.get_level_segments(level_name=self.source_level)
         target_level_segments = ts.hierarchical_structure.get_level_segments(level_name=self.target_level)
 
+        target_names = tuple(get_target_with_quantiles(columns=ts.columns))
```
Comment (Mr-Geekman) [inline etna/reconciliation/base.py:88]: We have already seen it in `_get_dataframe_level`.

File: etna/reconciliation/base.py [hunk 8]
```diff
@@ -83,18 +85,29 @@ def reconcile(self, ts: TSDataset) -> TSDataset:
...
+
         ts_reconciled = TSDataset(
             df=df_reconciled,
             freq=ts.freq,
             df_exog=ts.df_exog,
             known_future=ts.known_future,
             hierarchical_structure=ts.hierarchical_structure,
         )
+        if ts.target_components_names is not None:
+            ts_reconciled.add_target_components(target_components_df=target_components_df)
```
Comment (Mr-Geekman) [inline etna/reconciliation/base.py:112]: May be we should make `add_target_components` immute to `None` value.

File: tests/conftest.py [hunk 9]
```diff
@@ -683,26 +683,123 @@ def product_level_constant_forecast_w_quantiles(hierarchical_structure):
             "timestamp": ["2000-01-05", "2000-01-06"] * 4,
             "segment": ["a"] * 2 + ["b"] * 2 + ["c"] * 2 + ["d"] * 2,
             "target": [1, 1] + [2, 2] + [3, 3] + [4, 4],
-            "target_0.25": [1, 1] + [2, 2] + [3, 3] + [4, 4],
-            "target_0.75": [1, 1] + [2, 2] + [3, 3] + [4, 4],
-        }
+            "target_0.25": [1 / 2, 1 / 4] + [1, 1 / 2] + [2, 1] + [3, 2],
```
Comment (Mr-Geekman) [inline tests/conftest.py:686]: Why do we need this change?

File: tests/conftest.py [hunk 10]
```diff
@@ -683,26 +683,123 @@ def product_level_constant_forecast_w_quantiles(hierarchical_structure):
...
+        },
+        dtype=float,
     )
     df = TSDataset.to_dataset(df=df)
     ts = TSDataset(df=df, freq="D", hierarchical_structure=hierarchical_structure)
     return ts
 
 
+@pytest.fixture
+def product_level_constant_forecast_w_target_components(hierarchical_structure):
```
Comment (Mr-Geekman) [inline tests/conftest.py:697]: Both `with` and `without` starts with `w`))

File: tests/conftest.py [hunk 11]
```diff
@@ -683,26 +683,123 @@ def product_level_constant_forecast_w_quantiles(hierarchical_structure):
...
 def total_level_constant_forecast_w_quantiles(hierarchical_structure):
     df = pd.DataFrame(
         {
             "timestamp": ["2000-01-05", "2000-01-06"],
             "segment": ["total"] * 2,
             "target": [10, 10],
-            "target_0.25": [10, 10],
-            "target_0.75": [10, 10],
-        }
+            "target_0.25": [6.5, 3.75],
```
Comment (Mr-Geekman) [inline tests/conftest.py:772]: Why do we need this change?

File: tests/test_datasets/test_hierarchical_dataset.py [hunk 12]
```diff
@@ -471,21 +470,33 @@ def test_get_level_dataset_lower_level_error(simple_hierarchical_ts):
...
-    np.testing.assert_array_almost_equal(forecast.get_level_dataset(target_level=target_level).df.values, answer)
+def test_get_level_dataset_with_quantiles(
+    product_level_constant_forecast_w_quantiles, target_level, expected_dataframe_name, request
+):
+    expected_df = request.getfixturevalue(expected_dataframe_name).to_pandas()
+    reconciled_df = product_level_constant_forecast_w_quantiles.get_level_dataset(target_level=target_level).to_pandas()
+    pd.testing.assert_frame_equal(reconciled_df, expected_df)
 
 
-def test_get_level_dataset_pass_target_components_to_output(simple_hierarchical_ts):
```
Comment (Mr-Geekman) [inline tests/test_datasets/test_hierarchical_dataset.py:486]: Why do we remove this test?

File: tests/test_datasets/test_utils.py [hunk 13]
```diff
@@ -197,18 +197,21 @@ def test_get_target_with_quantiles(segments, columns, answer):
...
-    ts = product_level_simple_hierarchical_ts
-    answer = request.getfixturevalue(answer_name)
+@pytest.mark.parametrize(
+    "target_level, answer_name",
+    (("market", "market_level_constant_forecast_w_quantiles"), ("total", "total_level_constant_forecast_w_quantiles")),
+)
+def test_get_level_dataframe(product_level_constant_forecast_w_quantiles, target_level, answer_name, request):
+    ts = product_level_constant_forecast_w_quantiles
+    answer = request.getfixturevalue(answer_name).to_pandas()
     answer.index.freq = "D"
```
Comment (Mr-Geekman) [inline tests/test_datasets/test_utils.py:207]: Why is it like this? Where do we lose `freq`?

File: tests/test_reconciliation/test_base.py [hunk 14]
```diff
@@ -105,3 +120,26 @@ def test_reconcile(hierarchical_ts, source_level, target_level, mapping_matrix,
...
+    hierarchical_ts_with_target_components, source_level, target_level, mapping_matrix, request
+):
+    source_ts = hierarchical_ts_with_target_components.get_level_dataset(target_level=source_level)
+    expected_ts = hierarchical_ts_with_target_components.get_level_dataset(target_level=target_level)
+
+    reconciliator = DummyReconciliator(target_level=target_level, source_level=source_level)
+    reconciliator.mapping_matrix = request.getfixturevalue(mapping_matrix)
+    obtained_ts = reconciliator.reconcile(ts=source_ts)
+
+    assert obtained_ts.freq == expected_ts.freq
```
Comment (Mr-Geekman) [inline tests/test_reconciliation/test_base.py:139]: We should probably define `__eq__` in our `TSDataset`...

File: etna/datasets/tsdataset.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1136]: Not sure that is it possible as all the logic can't be placed inside one if

File: etna/datasets/tsdataset.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1116]: You mean create method which return dataframe with target, quantiles and target components?

File: etna/datasets/utils.py (hunk 5 above)
Comment (alex-hse-repository) [reply etna/datasets/utils.py:233]: It is a tricky point, we need segments to be ordered in the сertain order to perform the following matrix multiplication

File: etna/datasets/utils.py (hunk 6 above)
Comment (alex-hse-repository) [reply etna/datasets/utils.py:243]: We can, but here it is important that segments should be in the specific order

File: etna/reconciliation/base.py (hunk 8 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:112]: We can create private method which can handle `None`, but not public

File: tests/conftest.py (hunk 9 above)
Comment (alex-hse-repository) [reply tests/conftest.py:686]: To make sure that transformations are correct and we do't shuffle the columns values

File: tests/conftest.py (hunk 10 above)
Comment (alex-hse-repository) [reply tests/conftest.py:697]: Follow the naming of the existing fixtures)

Comment (alex-hse-repository) [reply tests/conftest.py:697]: Fixed

File: tests/conftest.py (hunk 11 above)
Comment (alex-hse-repository) [reply tests/conftest.py:772]: To make sure that transformations are correct and we don't shuffle the columns values

File: tests/test_datasets/test_hierarchical_dataset.py (hunk 12 above)
Comment (alex-hse-repository) [reply tests/test_datasets/test_hierarchical_dataset.py:486]: Replaced it with `test_get_level_dataset_with_target_components`

File: tests/test_datasets/test_utils.py (hunk 13 above)
Comment (alex-hse-repository) [reply tests/test_datasets/test_utils.py:207]: It is for `pd.testing`

File: tests/test_reconciliation/test_base.py (hunk 14 above)
Comment (alex-hse-repository) [reply tests/test_reconciliation/test_base.py:139]: Tried to do so earlier but there is a problem with comparing dtaframes

File: tests/test_datasets/test_utils.py (hunk 13 above)
Comment (alex-hse-repository) [reply tests/test_datasets/test_utils.py:207]: Will check the entire fixture

File: etna/datasets/tsdataset.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1122]: Create property `target_quantiles` in dataset

File: etna/datasets/tsdataset.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1142]: Change default value of `target_components_names`

File: etna/datasets/utils.py (hunk 5 above)
Comment (alex-hse-repository) [reply etna/datasets/utils.py:233]: Add comments in docstring

File: etna/reconciliation/base.py (hunk 7 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:88]: Fixed

File: etna/reconciliation/base.py (hunk 8 above)
Comment (alex-hse-repository) [reply etna/reconciliation/base.py:112]: Not sure it is useful

File: tests/test_datasets/test_utils.py (hunk 13 above)
Comment (alex-hse-repository) [reply tests/test_datasets/test_utils.py:207]: Fixed

File: tests/test_pipeline/test_pipeline.py [hunk 15]
```diff
@@ -116,7 +116,7 @@ def test_forecast_return_components(
     pipeline = Pipeline(model=model)
     pipeline.fit(example_tsds)
     forecast = pipeline.forecast(return_components=True)
-    assert forecast.target_components_names is not None
+    assert len(forecast.target_components_names) > 0
 
     taregt_components_df = TSDataset.to_flatten(forecast.get_target_components())
```
Comment (Mr-Geekman) [inline tests/test_pipeline/test_pipeline.py:121]: `taregt_components_df` -> `target_components_df`.

File: tests/test_pipeline/test_pipeline.py [hunk 16]
```diff
@@ -1201,7 +1201,7 @@ def test_predict_return_components(
     pipeline = Pipeline(model=model)
     pipeline.fit(example_tsds)
     forecast = pipeline.predict(ts=example_tsds, return_components=True)
-    assert forecast.target_components_names is not None
+    assert len(forecast.target_components_names) > 0
 
     taregt_components_df = TSDataset.to_flatten(forecast.get_target_components())
```
Comment (Mr-Geekman) [inline tests/test_pipeline/test_pipeline.py:1206]: Same about rename.

**Review by Mr-Geekman — CHANGES_REQUESTED** [review]

File: etna/datasets/tsdataset.py [hunk 17]
```diff
@@ -1134,7 +1143,9 @@ def get_level_dataset(self, target_level: str) -> "TSDataset":
             known_future=self.known_future,
             hierarchical_structure=self.hierarchical_structure,
         )
-        ts._target_components_names = self._target_components_names
+
+        if len(self.target_components_names) > 0:
+            ts.add_target_components(target_components_df=target_components_df)
```
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1148]: We have empty DataFrame with multi-index

File: etna/datasets/tsdataset.py [hunk 18]
```diff
@@ -1154,7 +1165,7 @@ def add_target_components(self, target_components_df: pd.DataFrame):
         ValueError:
             If components don't sum up to target
         """
-        if self._target_components_names is not None:
+        if len(self.target_components_names) > 0:
```
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1168]: I thing it is better to leave the current version, it is more readable

File: etna/datasets/tsdataset.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:1148]: I decided not to add the ability to pass empty dataframe in `add_target_components` as technically it is incorrect dataframe -- components don't sum up to target. It doesn't pass the first check in method `add_target_components` checking the components consistency between segments as there is not segments in empty datagrame

**Review by Mr-Geekman — APPROVED** [review]

**Review by brsnw250 — APPROVED** [review]

**Pull request merged**